### ISSUE TITLE
Support for attention masks in blocksparse layers

### ISSUE THREAD

**Issue by colehawkins** [facebookresearch/xformers#289]
# 🚀 Feature

Support for attention masks in blocksparse layers.

## Motivation

It would be nice.

## Pitch

We can pre-collect indices based on the sparse layout with this code when we run `create_triton_kernels`:

```python
            #set up preddetermined indices for gather operation during masking
            self.to_gather = []
            self.broadcast_to_gather = [] #use this if broadcasting same mask across heads
            num_col_blocks=self.layout.shape[-1]
            for idx, (h, i, j) in enumerate(zip(*self.layout.nonzero(as_tuple=True))):
                #assume mask is not 2D, otherwise very expensive
                self.to_gather.append(h*num_col_blocks+j)
                self.broadcast_to_gather.append(j)

            self.to_gather = torch.tensor(self.to_gather).long().to(device)
            self.broadcast_to_gather = torch.tensor(self.broadcast_to_gather).long().to(device)
```

Then we reshape and collect indices from the attention mask in the `forward`:
```python

            #apply masks
            if att_mask is not None:# and len(att_mask.nonzero)>0:
                #shape is either (B,1,1,S) or (B,nh,1,S)
                assert att_mask.shape[0]==q.shape[0] and att_mask.shape[1] in [q.shape[1],1] and att_mask.shape[-1]==q.shape[-2]
                #reshape into blocks
                block_att_mask = att_mask.reshape(att_mask.shape[0],att_mask.shape[1]*att_mask.shape[-1]//self.block_size,self.block_size)

                #gather based on predefined self.layout
                if att_mask.shape[1]==1:
                    block_att_mask = block_att_mask[:,self.broadcast_to_gather,:]
                else:
                    block_att_mask = block_att_mask[:,self.to_gather,:]

                #broadcast across the att_mask row dimension
                sparse_att_mat += block_att_mask.unsqueeze(-2)
```

## Alternatives

We could additionally support a 2D dense attention masking by calling `triton.testing.sparsify_tensor` but this is super slow.

## Additional context

We can probably support key_padding_mask as well. I am not sure of the standard shape for this, but the code should be similar.

I am using this locally and the speed seems fine. Happy to do a more thorough verification though.

**Comment by blefaudeux** [comment]
cc @ptillet :D

**Comment by blefaudeux** [comment]
hey @colehawkins, sorry for the delay, thanks for the issue and detailed suggestion, I think it's a great idea ! I think it could be something to upstream to triton directly actually, on the blocksparse side we just consume their op and they are PR-friendly. If it's not an option we could certainly fork the upstream kernel and host it here, but I would try to upstream it first and only use that as a fallback and this is a moving target, triton is under active development and blocksparse is a flag bearer for it. Thoughts @ptillet @colehawkins ? I don't have a lot of cycles myself but I can certainly try to support

**Comment by colehawkins** [comment]
@blefaudeux thanks for the feedback! I’m happy to proceed with either solution.

**Comment by blefaudeux** [comment]
> @blefaudeux Thanks for the nudge. Bumping this up the priority list. Triton softmax supports attention masks (or at least accepts the argument) so I'll try that first, and if I fail it'll be pytorch-only.

no pressure, I was just trying to triage a little !

**Comment by alexanderswerdlow** [comment]
Wanted to bump this since I'm looking to use a per-element attention mask. I saw that triton has since removed this as an argument. I'd be happy to help implement the feature if you had any suggestions.

**Comment by blefaudeux** [comment]
@alexanderswerdlow https://github.com/facebookresearch/xformers/pull/331 was the way to go I think, use the bias to nuke the attention on a per element basis. It's been some time, but top of mind this PR was almost there, we asked for minor convention changes to align with the other layers and that was pretty much it. I don't know whether @colehawkins is still interested in the topic ? Else I guess that you could take over, but I would sync with @fmassa who had plans on how to handle various attentions behind the same interface (using a dispatch to the proper op depending on the mask)

On a related note, xformers has been tied to a relatively old version of triton for a while, for now you would need to stick to this version for the unit tests to pass. I'm planning on submitting a PR moving everything to "modern" triton2 (post major refactor that they did a couple of months ago) in the coming days, that may help

**Comment by colehawkins** [comment]
@blefaudeux Unfortunately I'm very limited on time right now and expecting that to continue for 1-2 months. At that point I'm likely to return. Happy for someone else to take over. @alexanderswerdlow
